legacy/firmware/config.c

```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
   return sectrue == ret;
 }
 
-const uint8_t *session_getSessionId(void) {
-  if (!sessionIdCached) {
-    random_buffer(sessionId, 32);
+uint8_t session_findLRU(void) {
+  uint8_t lowest_index = MAX_SESSIONS_COUNT;
+  uint32_t lowest = sessionLRUCounter;
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage < lowest) {
+      lowest = sessionsCache[i].usage;
+      lowest_index = i;
+    }
+  }
+  // TODO: abort if lowest_index == MAX_SESSIONS_COUNT?
+  // TODO: it should never happen, so it is more of "internal" check
+  return lowest_index;
+}
+
+uint8_t session_findSession(const uint8_t *sessionId) {
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage != 0) {
+      if (memcmp(sessionsCache[i].id, sessionId, 32) == 0) {  // session found
+        return i;
+      }
+    }
+  }
+  return MAX_SESSIONS_COUNT;
+}
+
+uint8_t session_findEmpty(void) {
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage == 0) {
+      return i;
+    }
+  }
+  return MAX_SESSIONS_COUNT;
+}
+
+uint8_t *session_startSession(const uint8_t *received_session_id) {
+  int session_index = MAX_SESSIONS_COUNT;
+  if (received_session_id != NULL) {
+    session_index = session_findSession(received_session_id);
   }
-  sessionIdCached = sectrue;
-  return sessionId;
+  if (session_index != MAX_SESSIONS_COUNT) {
+    // session found in currently stored sessions
+    sessionLRUCounter++;
+    sessionsCache[session_index].usage = sessionLRUCounter;
+    activeSessionCache = sessionsCache + session_index;
+    return activeSessionCache->id;
+  } else {
+    // if not found or session_id not provided try to find an empty one
+    session_index = session_findEmpty();
+  }
+  if (session_index == MAX_SESSIONS_COUNT) {
+    // if still no success, we need to clear the LRU item
+    session_index = session_findLRU();
+  }
+
+  // now it is guaranteed `session_index` is the place to start a new session
+  sessionLRUCounter++;
+  session_clearCache(sessionsCache + session_index);
+
+  sessionsCache[session_index].usage = sessionLRUCounter;
+  random_buffer(sessionsCache[session_index].id, 32);
+
+  activeSessionCache = sessionsCache + session_index;
+  return activeSessionCache->id;
```

Comment:
I find this a bit hard to follow. I think it can be written more cleanly:
```
uint8_t *session_startSession(const uint8_t *received_session_id) {
  int session_index = MAX_SESSIONS_COUNT;

  if (received_session_id != NULL) {
    session_index = session_findSession(received_session_id);
  }

  if (session_index == MAX_SESSIONS_COUNT) {
    // Session not found in cache. Use the least recently used session in the cache.
    session_index = session_findLRU();
    session_clearCache(sessionsCache + session_index);
    random_buffer(sessionsCache[session_index].id, 32);
  }

  sessionLRUCounter++;
  sessionsCache[session_index].usage = sessionLRUCounter;
  activeSessionCache = sessionsCache + session_index;
  return activeSessionCache->id;
}
```

Reply:
Yes, that looks better.